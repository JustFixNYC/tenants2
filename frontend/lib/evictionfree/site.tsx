import loadable from "@loadable/component";
import React, { useContext } from "react";
import Headroom from "react-headroom";
import { Link, Route, useLocation } from "react-router-dom";
import type { AppSiteProps } from "../app";
import { createLinguiCatalogLoader, li18n } from "../i18n-lingui";
import { LoadingOverlayManager } from "../networking/loading-page";
import { EvictionFreeRouteComponent } from "./routes";
import {
  EvictionFreeRoutes as Routes,
  getEvictionFreeRoutesForPrimaryPages,
  useEvictionFreeUnsupportedLocale,
} from "./route-info";
import Navbar, { NavbarDropdown } from "../ui/navbar";
import { AppContext, getGlobalAppServerInfo } from "../app-context";
import { t, Trans } from "@lingui/macro";
import { LANGUAGE_NAMES, SwitchLanguage } from "../ui/language-toggle";
import classnames from "classnames";
import { EvictionFreeFooter } from "./components/footer";
import { EvictionFreeHelmet } from "./components/helmet";
import i18n from "../i18n";
import {
  EvictionFreeUnsupportedLocaleChoices,
  getEvictionFreeUnsupportedLocaleChoiceLabels,
} from "../../../common-data/evictionfree-unsupported-locale-choices";
import { SwitchToUnsupportedLanguage } from "./unsupported-locale";
import { Modal } from "../ui/modal";
import { LocalizedOutboundLink } from "../ui/localized-outbound-link";

export const EvictionFreeLinguiI18n = createLinguiCatalogLoader({
  en: loadable.lib(
    () => import("../../../locales/en/evictionfree.chunk") as any
  ),
  es: loadable.lib(
    () => import("../../../locales/es/evictionfree.chunk") as any
  ),
});

function useIsPrimaryPage() {
  const location = useLocation();
  return getEvictionFreeRoutesForPrimaryPages().includes(location.pathname);
}

const EvictionFreeBrand: React.FC<{}> = () => {
  const isPrimaryPage = useIsPrimaryPage();
  return (
    <Link className="navbar-item" to={Routes.locale.home}>
      <span
        className={classnames(
          "jf-evictionfree-logo",
          isPrimaryPage ? "has-text-info" : "has-text-white"
        )}
      >
        Eviction
        <br />
        Free NY
      </span>
    </Link>
  );
};

export const EvictionFreeLanguageDropdown: React.FC<{}> = () => {
  const { server } = useContext(AppContext);
  const locales = server.enabledLocales;
  const activeLocale = i18n.locale;
  const unsupportedLocale = useEvictionFreeUnsupportedLocale();
  const languageName = unsupportedLocale
    ? getEvictionFreeUnsupportedLocaleChoiceLabels()[unsupportedLocale]
    : LANGUAGE_NAMES[activeLocale];

  return (
    <NavbarDropdown id="locale" label={languageName}>
      {locales
        .filter((locale) =>
          !unsupportedLocale ? locale !== activeLocale : true
        )
        .map((locale) => (
          <SwitchLanguage
            key={locale}
            locale={locale}
            linkToCurrentLocale={!!unsupportedLocale}
            className="navbar-item"
          />
        ))}
      {EvictionFreeUnsupportedLocaleChoices.filter(
        (locale) => locale !== unsupportedLocale
      ).map((locale) => (
        <SwitchToUnsupportedLanguage
          locale={locale}
          className="navbar-item"
          key={locale}
        />
      ))}
    </NavbarDropdown>
  );
};

const EvictionFreeBuildMyDeclarationLink: React.FC<{}> = () => {
  const isPrimaryPage = useIsPrimaryPage();
  return (
    <>
      <div className="navbar-item is-hidden-touch">
        <Link
          className={classnames(
            "button",
            isPrimaryPage ? "is-primary" : "is-info is-inverted is-outlined"
          )}
          to={Routes.locale.declaration.latestStep}
        >
          <Trans>Fill out my form</Trans>
        </Link>
      </div>
      <Link
        className="navbar-item is-hidden-desktop"
        to={Routes.locale.declaration.latestStep}
      >
        <Trans>Fill out my form</Trans>
      </Link>
    </>
  );
};

const EvictionFreeMenuItems: React.FC<{}> = () => {
  const { session } = useContext(AppContext);
  const siteIsActive = !getGlobalAppServerInfo().isEfnySuspended;
  return (
    <>
      {siteIsActive && (
        <>
          <Link className="navbar-item" to={Routes.locale.faqs}>
            <Trans>Faqs</Trans>
          </Link>
          <Link className="navbar-item" to={Routes.locale.about}>
            <Trans>About</Trans>
          </Link>
        </>
      )}
      {session.phoneNumber ? (
        <Link className="navbar-item" to={Routes.locale.logout}>
          <Trans>Log out</Trans>
        </Link>
      ) : (
        <Link className="navbar-item" to={Routes.locale.login}>
          <Trans>Log in</Trans>
        </Link>
      )}
      <EvictionFreeLanguageDropdown />
      {siteIsActive && <EvictionFreeBuildMyDeclarationLink />}
    </>
  );
};

export const EvictionFreeSuspendedModal = () => (
  <Modal
    title={li18n._(t`Eviction Free NY has been suspended`)}
    onCloseGoTo={1}
  >
    <div className="jf-is-scrollable-if-too-tall has-text-centered">
      <p>
        <Trans id="evictionfree.toolSuspensionMessage">
          As of January 15, 2022, tenants in New York State are no longer
          protected from eviction after submitting a declaration of hardship.
          However, you still have other options to protect yourself and your
          neighbors!
        </Trans>
      </p>
      <br />
      <button className="button is-primary jf-build-my-declaration-btn jf-is-extra-wide">
        <LocalizedOutboundLink
          hrefs={{
            en:
              "https://www.righttocounselnyc.org/eviction_protections_during_covid",
            es:
              "https://www.righttocounselnyc.org/protecciones_contra_desalojos",
          }}
        >
          Learn More
        </LocalizedOutboundLink>
      </button>
      <br />
      <br />
      <p className="is-size-7">
        <Trans>
          Already filled out a hardship declaration form?{" "}
          <Link to={Routes.locale.login}>Log in here</Link>
        </Trans>
      </p>
    </div>
  </Modal>
);

const EvictionFreeSite = React.forwardRef<HTMLDivElement, AppSiteProps>(
  (props, ref) => {
    const isPrimaryPage = useIsPrimaryPage();
    const isHomepage = useLocation().pathname === Routes.locale.home;

    return (
      <EvictionFreeLinguiI18n>
        <EvictionFreeSuspendedModal />
        <section
          className={classnames(
            isPrimaryPage
              ? "jf-above-footer-content"
              : "jf-norent-internal-above-footer-content"
          )}
        >
          <span
            className={classnames(
              isPrimaryPage && "jf-white-navbar",
              isHomepage && "jf-evictionfree-homepage-navbar"
            )}
          >
            {/* NOTE: We wanted to originally wrap ALL page navbars in this <Headroom> component,
            but unfortunately we noticed that this library interacts strangely with our page transitions
            between declaration builder steps and messes up the animation, so we are only including it 
            on the primary pages for now. */}
            {isPrimaryPage ? (
              <Headroom>
                <Navbar
                  menuItemsComponent={EvictionFreeMenuItems}
                  brandComponent={EvictionFreeBrand}
                />
              </Headroom>
            ) : (
              <Navbar
                menuItemsComponent={EvictionFreeMenuItems}
                brandComponent={EvictionFreeBrand}
              />
            )}
            <EvictionFreeHelmet />
          </span>
          {!isPrimaryPage && (
            <div className="jf-block-of-color-in-background" />
          )}
          <div
            ref={ref}
            data-jf-is-noninteractive
            tabIndex={-1}
            className={classnames(
              !isPrimaryPage && "box jf-norent-builder-page"
            )}
          >
            <LoadingOverlayManager>
              <Route component={EvictionFreeRouteComponent} />
            </LoadingOverlayManager>
          </div>
        </section>
        <EvictionFreeFooter />
      </EvictionFreeLinguiI18n>
    );
  }
);

export default EvictionFreeSite;
