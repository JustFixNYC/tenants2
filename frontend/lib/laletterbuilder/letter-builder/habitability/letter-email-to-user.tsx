import React from "react";

import { asEmailStaticPage } from "../../../static-page/email-static-page";
import { li18n } from "../../../i18n-lingui";
import { t, Trans } from "@lingui/macro";
import { USPS_TRACKING_URL_PREFIX } from "../../../../../common-data/loc";
import { HtmlEmail } from "../../../static-page/html-email";
import { TransformSession } from "../../../util/transform-session";
import { AllSessionInfo } from "../../../queries/AllSessionInfo";

type EmailBodyProps = {
  firstName: string;
  trackingNumber?: string;
  isInLosAngeles: boolean;
};

const EmailBody: React.FC<EmailBodyProps> = (props) => (
  <>
    <Trans>
      <p>Hello {props.firstName},</p>
      <p>
        You've sent your Notice To Repair letter. Attached to this email is a
        PDF copy for your records.
      </p>
    </Trans>
    {props.trackingNumber && (
      <p>
        <Trans>
          You can also track the delivery of your letter using USPS Tracking:
        </Trans>{" "}
        <a
          data-jf-show-href-only-in-plaintext
          href={USPS_TRACKING_URL_PREFIX + props.trackingNumber}
        >
          {props.trackingNumber}
        </a>
      </p>
    )}
    {props.isInLosAngeles ? (
      <>
        <p>
          <Trans>
            You can contact Strategic Actions for a Just Economy (SAJE) - a
            501c3 non-profit organization in South Los Angeles.
          </Trans>
        </p>
        <blockquote>
          <em>
            <Trans id="norent.sajeBlockQuote">
              Since 1996 SAJE has been a force for economic justice in our
              community focusing on tenant rights, healthy housing, and
              equitable development. SAJE believes that the fate of city
              neighborhoods should be decided by those who dwell there, and
              convenes with other organizations to ensure this occurs in a
              manner that is fair, replicable, and sustainable. Housing is a
              human right.
            </Trans>
          </em>
        </blockquote>
        <p>
          <strong>
            <Trans id="norent.sajePhoneCalls">
              Strategic Actions for a Just Economy (SAJE) is available for phone
              calls at (213) 745-9961, Monday-Friday from 10:00am-6:00pm.
            </Trans>
          </strong>
        </p>
      </>
    ) : (
      <></>
    )}
  </>
);

export const HabitabilityLetterEmailToUserBody: React.FC<{}> = () => {
  return (
    <TransformSession
      transformer={(session: AllSessionInfo): EmailBodyProps | null => {
        const { firstName, onboardingInfo } = session;
        if (!(onboardingInfo && firstName)) return null;
        return {
          trackingNumber: session.habitabilityLatestLetter?.trackingNumber,
          isInLosAngeles: !!onboardingInfo.isInLosAngeles,
          firstName,
        };
      }}
    >
      {(bodyProps) => <EmailBody {...bodyProps} />}
    </TransformSession>
  );
};

export const HabitabilityLetterEmailToUserStaticPage = asEmailStaticPage(
  (props) => (
    <HtmlEmail
      subject={li18n._(
        t`Your Notice to Repair letter and important next steps`
      )}
    >
      <HabitabilityLetterEmailToUserBody />
    </HtmlEmail>
  )
);
