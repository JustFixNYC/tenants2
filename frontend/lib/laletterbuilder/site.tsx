import classnames from "classnames";
import React, { useContext } from "react";
import { useLocation, Route, Link } from "react-router-dom";

import { Trans } from "@lingui/macro";
import loadable from "@loadable/component";

import { AppContext } from "../app-context";
import { createLinguiCatalogLoader } from "../i18n-lingui";
import { LoadingOverlayManager } from "../networking/loading-page";
import { NavbarLanguageDropdown } from "../ui/language-toggle";
import Navbar from "../ui/navbar";
import { LaLetterBuilderFooter } from "./components/footer";
import {
  LaLetterBuilderRoutes as Routes,
  getLaLetterBuilderRoutesForPrimaryPages,
} from "./route-info";
import { LaLetterBuilderRouteComponent } from "./routes";

import type { AppSiteProps } from "../app";

export const LaLetterBuilderLinguiI18n = createLinguiCatalogLoader({
  en: loadable.lib(
    () => import("../../../locales/en/laletterbuilder.chunk") as any
  ),
  es: loadable.lib(
    () => import("../../../locales/es/laletterbuilder.chunk") as any
  ),
});

function useIsPrimaryPage() {
  const location = useLocation();
  return getLaLetterBuilderRoutesForPrimaryPages().includes(location.pathname);
}

const LaLetterBuilderBrand: React.FC<{}> = () => {
  return (
    <Link className="navbar-item" to={Routes.locale.home}>
      <span className="jf-laletterbuilder-logo">LA Letter Builder</span>
    </Link>
  );
};

const LaLetterBuilderMenuItems: React.FC<{}> = () => {
  const { session } = useContext(AppContext);
  return (
    <>
      <Link className="navbar-item" to={Routes.locale.letter.latestStep}>
        Build my letter
      </Link>
      {session.phoneNumber ? (
        <Link className="navbar-item" to={Routes.locale.logout}>
          <Trans>Log out</Trans>
        </Link>
      ) : (
        <Link className="navbar-item" to={Routes.locale.letter.phoneNumber}>
          <Trans>Log in</Trans>
        </Link>
      )}
      <NavbarLanguageDropdown />
    </>
  );
};

const LaLetterBuilderSite = React.forwardRef<HTMLDivElement, AppSiteProps>(
  (props, ref) => {
    const isPrimaryPage = useIsPrimaryPage();

    return (
      <LaLetterBuilderLinguiI18n>
        <section
          className={classnames(
            isPrimaryPage
              ? "jf-above-footer-content"
              : "jf-norent-internal-above-footer-content"
          )}
        >
          <span className={classnames(isPrimaryPage && "jf-white-navbar")}>
            <Navbar
              menuItemsComponent={LaLetterBuilderMenuItems}
              brandComponent={LaLetterBuilderBrand}
            />
          </span>
          {!isPrimaryPage && (
            <div className="jf-block-of-color-in-background" />
          )}
          <div
            ref={ref}
            data-jf-is-noninteractive
            tabIndex={-1}
            className={classnames(
              !isPrimaryPage && "box jf-norent-builder-page"
            )}
          >
            <LoadingOverlayManager>
              <Route component={LaLetterBuilderRouteComponent} />
            </LoadingOverlayManager>
          </div>
        </section>
        <LaLetterBuilderFooter />
      </LaLetterBuilderLinguiI18n>
    );
  }
);

export default LaLetterBuilderSite;
