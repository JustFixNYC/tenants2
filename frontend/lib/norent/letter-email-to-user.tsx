import React from "react";

import { useContext } from "react";
import { AppContext } from "../app-context";
import {
  EmailSubject,
  asEmailStaticPage,
} from "../static-page/email-static-page";
import { NorentRoutes } from "./routes";
import { li18n } from "../i18n-lingui";
import { t, Trans } from "@lingui/macro";
import { USPS_TRACKING_URL_PREFIX } from "../../../common-data/loc.json";

export const NorentLetterEmailToUser: React.FC<{}> = () => {
  const { session, server } = useContext(AppContext);
  const letter = session.norentLatestLetter;
  const faqURL = `${server.originURL}${NorentRoutes.locale.faqs}`;

  return (
    <>
      <EmailSubject value={li18n._(t`Here's a copy of your NoRent letter`)} />
      <Trans>
        <p>Hello {session.firstName},</p>
        <p>
          You've sent your NoRent letter. Attached to this email is a PDF copy
          for your records.
        </p>
      </Trans>
      {letter?.trackingNumber && <p>
        <Trans>
          You can also track the delivery of your letter using USPS Tracking:
      </Trans>{" "}
        {USPS_TRACKING_URL_PREFIX + letter.trackingNumber}
      </p>}
      <p>
        <Trans>
          To learn more about what to do next, check out our FAQ page: {faqURL}
        </Trans>
      </p>
    </>
  );
};

export const NorentLetterEmailToUserStaticPage = asEmailStaticPage(
  NorentLetterEmailToUser
);
