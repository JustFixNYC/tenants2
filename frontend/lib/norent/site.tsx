import React, { useContext } from "react";
import { AppSiteProps } from "../app";
import { NorentRoutes as Routes } from "./routes";
import { RouteComponentProps, Switch, Route, Link } from "react-router-dom";
import { NotFound } from "../pages/not-found";
import { NorentHomePage } from "./homepage";
import {
  LoadingPage,
  friendlyLoad,
  LoadingOverlayManager,
} from "../networking/loading-page";
import loadable from "@loadable/component";
import classnames from "classnames";
import { NorentLetterRoutes } from "./letter-builder";
import { AppContext } from "../app-context";
import { NorentFooter } from "./components/footer";
import {
  NorentLetterForUserStaticPage,
  NorentSampleLetterSamplePage,
} from "./letter-content";
import Navbar from "../ui/navbar";
import { createLetterStaticPageRoutes } from "../static-page/routes";
import { NorentFaqsPage } from "./faqs";
import { NorentAboutPage } from "./about";
import { NorentAboutYourLetterPage } from "./about-your-letter";

const ROUTES_FOR_PRIMARY_PAGES = [
  Routes.locale.home,
  Routes.locale.info,
  Routes.locale.faqs,
  Routes.locale.aboutLetter,
];

const LoadableDevRoutes = loadable(() => friendlyLoad(import("../dev/dev")), {
  fallback: <LoadingPage />,
});

const NorentRoute: React.FC<RouteComponentProps> = (props) => {
  const { location } = props;
  if (!Routes.routeMap.exists(location.pathname)) {
    return NotFound(props);
  }
  return (
    <Switch location={location}>
      <Route path={Routes.locale.home} exact component={NorentHomePage} />
      <Route path={Routes.locale.faqs} exact component={NorentFaqsPage} />
      <Route path={Routes.locale.info} exact component={NorentAboutPage} />
      <Route
        path={Routes.locale.aboutLetter}
        exact
        component={NorentAboutYourLetterPage}
      />
      <Route
        path={Routes.locale.letter.prefix}
        component={NorentLetterRoutes}
      />
      {createLetterStaticPageRoutes(Routes.locale.letterContent, (isPdf) => (
        <NorentLetterForUserStaticPage isPdf={isPdf} />
      ))}
      {createLetterStaticPageRoutes(
        Routes.locale.sampleLetterContent,
        (isPdf) => (
          <NorentSampleLetterSamplePage isPdf={isPdf} />
        )
      )}
      <Route path={Routes.dev.prefix} component={LoadableDevRoutes} />
      <Route component={NotFound} />
    </Switch>
  );
};

const NorentMenuItems: React.FC<{}> = () => {
  const { session } = useContext(AppContext);
  return (
    <>
      <Link className="navbar-item" to={Routes.locale.aboutLetter}>
        About your letter
      </Link>
      <Link className="navbar-item" to={Routes.locale.faqs}>
        Faqs
      </Link>
      <Link className="navbar-item" to={Routes.locale.info}>
        Information
      </Link>
      {session.phoneNumber ? (
        // These are placeholders just to show styling.
        // Will replace when we have site scaffolding ready.
        <Link className="navbar-item" to={Routes.locale.home}>
          Log out
        </Link>
      ) : (
        <Link className="navbar-item" to={Routes.locale.home}>
          Log in
        </Link>
      )}
    </>
  );
};

const NorentSite = React.forwardRef<HTMLDivElement, AppSiteProps>(
  (props, ref) => {
    const isPrimaryPage = ROUTES_FOR_PRIMARY_PAGES.includes(
      props.location.pathname
    );
    return (
      <>
        <section
          className={classnames(
            "section",
            "jf-above-footer-content",
            isPrimaryPage && "is-paddingless"
          )}
        >
          <Navbar menuItemsComponent={NorentMenuItems} />
          <div
            className={classnames(!isPrimaryPage && "container")}
            ref={ref}
            data-jf-is-noninteractive
            tabIndex={-1}
          >
            <LoadingOverlayManager>
              <Route component={NorentRoute} />
            </LoadingOverlayManager>
          </div>
        </section>
        <NorentFooter />
      </>
    );
  }
);

export default NorentSite;
