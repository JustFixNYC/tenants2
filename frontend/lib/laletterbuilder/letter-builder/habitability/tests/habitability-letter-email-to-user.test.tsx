import React from "react";
import { AppTesterPal } from "../../../../tests/app-tester-pal";
import { HabitabilityLetterEmailToUserBody } from "../letter-email-to-user";
import { LaLetterBuilderLinguiI18n } from "../../../site";
import { preloadLingui } from "../../../../tests/lingui-preloader";
import { newSb } from "../../../../tests/session-builder";

beforeAll(preloadLingui(LaLetterBuilderLinguiI18n));

describe("HabitabilityLetterEmailToUserBody", () => {
  it("works with LA users", () => {
    const pal = new AppTesterPal(<HabitabilityLetterEmailToUserBody />, {
      session: newSb().withLoggedInLosAngelesUser().withMailedNorentLetter()
        .value,
    });
    expect(pal.rr.container).toMatchSnapshot();
  });

  it("works with SF users", () => {
    const pal = new AppTesterPal(<HabitabilityLetterEmailToUserBody />, {
      session: newSb().withLoggedInSanFranciscoUser().withMailedNorentLetter()
        .value,
    });
    expect(pal.rr.container).toMatchSnapshot();
  });
});
