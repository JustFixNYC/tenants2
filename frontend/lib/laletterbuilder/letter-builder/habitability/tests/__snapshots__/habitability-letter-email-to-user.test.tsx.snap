// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HabitabilityLetterEmailToUserBody works with LA users 1`] = `
<div>
  <p>
    Hello Boop,
  </p>
  <p>
    You've sent your Notice To Repair letter. Attached to this email is a PDF copy for your records.
  </p>
  <p>
    You can contact Strategic Actions for a Just Economy (SAJE) - a 501c3 non-profit organization in South Los Angeles.
  </p>
  <blockquote>
    <em>
      Since 1996 SAJE has been a force for economic justice in our community focusing on tenant rights, healthy housing, and equitable development. SAJE believes that the fate of city neighborhoods should be decided by those who dwell there, and convenes with other organizations to ensure this occurs in a manner that is fair, replicable, and sustainable. Housing is a human right.
    </em>
  </blockquote>
  <p>
    <strong>
      Strategic Actions for a Just Economy (SAJE) is available for phone calls at (213) 745-9961, Monday-Friday from 10:00am-6:00pm.
    </strong>
  </p>
</div>
`;

exports[`HabitabilityLetterEmailToUserBody works with SF users 1`] = `
<div>
  <p>
    Hello Boop,
  </p>
  <p>
    You've sent your Notice To Repair letter. Attached to this email is a PDF copy for your records.
  </p>
</div>
`;
